### PR TITLE
chore: replace User.sol for use of prank()

### PR DESCRIPTION
**Pull request by fedealconada**
## Motivation

Have tests up to date with foundry's features

## Solution

Remove `User.sol` (which was a contract used to impersonate accounts) and replace it for the use of `prank()` or `startPrank()`/`stopPrank()`.

## Implementer Checklist

- [ ]  [FIRST TIME ONLY] Review the [Solcurity](https://github.com/Rari-Capital/solcurity) standard
- [ ]  [OPTIONAL] Create a reference implementation (in python or JS)
- [x]  Create a short "Motivation" section for the PR
- [ ]  Write a spec for the feature and put it in the PR description – basic function names and expected state transitions is OK
- [ ]  Check all of the new revert paths with concrete tests
- [ ]  Check all of the new storage slot writes with concrete tests – do not make the passing test case the trivial case!
- [ ]  Go line-by-line through the new code and ensure it has all been covered by concrete tests – since we don’t have a coverage engine for foundry yet, you can imagine how one might check that each different code path is covered
- [ ]  Simplify the implementation and spend some time trying to minimize gas costs
- [ ]  Write fuzz tests for the feature – try everything to break the implementation (is this function monotonically in/decreasing, should it always be less than something else, etc)
- [ ]  Capture bugs discovered during the above steps in concrete tests (regression tests)
- [ ]  Add integration tests – fuzz or concrete tests that test how the new code affects the entire system, either locally or with a mainnet fork
- [ ]  [OPTIONAL] Integrate the feature into the deployment scripts
- [ ]  [OPTIONAL] Add new sanity checks to the deployment scripts
- [ ]  Get the PR reviewed by at least two people

### CODE REVIEW

File: pkg/core/src/tests/YT.t.sol [hunk 1]
```diff
@@ -66,20 +72,22 @@ contract Yield is TestHelper {
...
-        uint256 aytBalanceAfter = ERC20(yt).balanceOf(address(alice));
-        uint256 bytBalanceAfter = ERC20(yt).balanceOf(address(bob));
-        uint256 tBalanceAfter = target.balanceOf(address(bob));
+        uint256 lscale = divider.lscales(address(adapter), maturity, bob);
+        uint256 aytBalanceBefore = ERC20(yt).balanceOf(alice);
+        uint256 bytBalanceBefore = ERC20(yt).balanceOf(bob);
+        uint256 tBalanceBefore = target.balanceOf(bob);
+        hevm.prank(bob);
+        ERC20(yt).approve(alice, type(uint256).max);
+        Token(yt).transferFrom(bob, alice, bytBalanceBefore);
```
Comment (jparklev) [inline pkg/core/src/tests/YT.t.sol:87]: It looks like we approve the `alice` address but initiate the `transferFrom` from the test contract address?

File: pkg/core/src/tests/test-helpers/TestHelper.sol [hunk 2]
```diff
@@ -226,43 +227,42 @@ contract TestHelper is DSTest {
...
-        updateUser(user, target, stake, amt);
+        alice = address(this); // alice is the default user
+        initUser(alice, target, MAX_TARGET);
+        initUser(bob, target, MAX_TARGET);
+        initUser(jim, target, MAX_TARGET);
     }
 
-    function updateUser(
-        User user,
+    function initUser(
```
Comment (jparklev) [inline pkg/core/src/tests/test-helpers/TestHelper.sol:239]: Could [`label` the addresses](https://github.com/foundry-rs/forge-std/blob/78adc23f035d9ed4f6fedda5cccf80b1d4a9b9a6/src/Vm.sol#L91) as well so it's easier to tell which address is which in the trace

File: pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol [hunk 3]
```diff
@@ -137,9 +137,9 @@ contract MockAdapter is BaseAdapter, Crop {
         tBal = Divider(divider).combine(address(this), maturity, uBal);
     }
 
-    function doRedeemPrincipal(uint256 maturity, uint256 uBal) external {
-        Divider(divider).redeem(address(this), maturity, uBal);
-    }
+    // function doRedeemPrincipal(uint256 maturity, uint256 uBal) external {
```
Comment (jparklev) [inline pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol]: Good to delete these?

**Review by jparklev — APPROVED** [review]
✨ quality work ✨

LGTM. How does it feel to you to do it this way? Also are you still running into the issue with `prank` not always working that you'd mentioned to me before?

File: pkg/core/src/tests/YT.t.sol (hunk 1 above)
Comment (fedealconada) [reply pkg/core/src/tests/YT.t.sol:87]: yes, this is because in the `setUp()` i'm setting alice as the default address `alice = address(this)`

**Comment by fedealconada** [comment]
> LGTM. How does it feel to you to do it this way? 

it feels good but i have to say I kind of like what we had to be honest, like calling things like `alice.doIssue` it was very clear who was performing the action but it might be that I just got used to that. And I think we are better following foundry's features rather than having the User.sol workaround.

> Also are you still running into the issue with prank not always working that you'd mentioned to me before?

No! It was actually that I was not seeing that I was doing another call haha. As it was like "inline" i was not seeing it! Something like:
```
hevm.prank(alice);
divider.issue(address(adapter), maturity, ERC20(target).balanceOf(alice));
```
(note that `balanceOf` call)

File: pkg/core/src/tests/test-helpers/TestHelper.sol (hunk 2 above)
Comment (fedealconada) [reply pkg/core/src/tests/test-helpers/TestHelper.sol:239]: added!

**Comment by jparklev** [comment]
> And I think we are better following foundry's features rather than having the User.sol workaround.

Yep, feel similarly and ultimately agree

> No! It was actually that I was not seeing that I was doing another call haha

Ahh classic, well glad you caught it. This PR LGTM now :)

File: pkg/core/src/tests/adapters/CropsAdapter.t.sol [hunk 4]
```diff
@@ -39,9 +40,10 @@ contract CropsAdapters is TestHelper {
         cropsAdapter = MockCropsAdapter(f);
         divider.setGuard(address(cropsAdapter), 10 * 2**128);
 
-        updateUser(alice, aTarget, stake, MAX_TARGET);
-        updateUser(bob, aTarget, stake, MAX_TARGET);
-        updateUser(jim, aTarget, stake, MAX_TARGET);
+        alice = alice;
```
Comment (abylaw) [inline pkg/core/src/tests/adapters/CropsAdapter.t.sol]: nit: what's this for? :)

File: pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol [hunk 5]
```diff
@@ -423,7 +419,7 @@ contract MockBaseAdapter is BaseAdapter {
         return 1e18;
     }
 
-    function doSetAdapter(Divider d, address _adapter) public {
-        d.setAdapter(_adapter, true);
-    }
+    // function doSetAdapter(Divider d, address _adapter) public {
+    //     d.setAdapter(_adapter, true);
+    // }
```
Comment (abylaw) [inline pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol]: delete these too?

**Review by abylaw — APPROVED** [review]
thanks for doing this :)

> And I think we are better following foundry's features rather than having the User.sol workaround.

+1

File: pkg/core/src/tests/adapters/CropsAdapter.t.sol (hunk 4 above)
Comment (fedealconada) [reply pkg/core/src/tests/adapters/CropsAdapter.t.sol]: removing! :)

File: pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol (hunk 5 above)
Comment (fedealconada) [reply pkg/core/src/tests/test-helpers/mocks/MockAdapter.sol]: yes! ty!

**Pull request merged**